### PR TITLE
feat(refactor): Add `Move Selection To` command

### PR DESCRIPTION
**Pull request by hikchoi**
# feat(refactor): Add `Move Selection To` command

This PR:
- refactors editor related utils
- refactors common content level operations to reusable utilities
- fixes `selectionExtract` so that it works on existing notes as well
- Adds new command `Dendron: Move To`
- ~~(pending) removes all edit builder operations to be explicit engine API calls so that note changes can be tracked within `NoteLookupCommand` instead of the watcher~~
  - dropped since this was a very roundabout way to track changes and was unnecessary

# Pull Request Checklist
### Basics

- [x] code should follow [Code Conventions](dev.process.code)
- [x] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](dev.process.code.best-practices).
- [x] sticking to existing conventions instead of creating new ones
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended
- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [x] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)

### Basics
- [x] if you are adding analytics related changes, make sure the [Telemetry](https://wiki.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated

### Extended
- [x] can we track the performance of this change to know if it is _successful_?
  - eg: [see usage for export pod](https://github.com/dendronhq/dendron/pull/2190#pullrequestreview-855715612)

### Basics

- [x] [Write Tests](dev.process.qa.test) 
- [x] [Confirm existing tests pass](dev.process.qa.test)
- [x] [Confirm manual testing](dev.process.qa.test) 
- [x] Common cases tested
- [x] 1-2 Edge cases tested
- [~] If your tests changes an existing snapshot, snapshots have been [updated](dev.process.qa.test)

### Extended
- [~] If you are adding a new language feature (graphically visible in VS Code/preview/publishing), an example is included in the [test workspace](dev.ref.test-workspace)
- CSS
  - [~] display is correct for following dimensions
    - [~] sm: screen ≥ 576px, eg. iphonex, (375x812)
    - [~] lg: screen ≥ 992px
    - [~] xxl: screen ≥ 1600px eg. mac (1600x900)
  - [~] display is correct for following browsers (across the various dimensions)
    - [~] safari
    - [~] firefox
    - [~] chrome

## Docs
- [x] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR
- [~] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](dev) or [Packages](pkg)

### Extended
- [~]  is this a developer BREAKING change? if another person cloning from this branch will need to adjust their dependencies or mental model of the architecture, then it is. if this is the case, make sure this is communicated according to [Close Loop](dev.process.close-loop)
  - eg: [breaking dev change due to new dependency](https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330)

### CODE REVIEW

**Comment by hikchoi** [comment]
pending:

- [ ] add tests
- [ ] add docs
- [ ] update airtable with new event props

**Comment by hikchoi** [comment]
docs: https://github.com/dendronhq/dendron-site/pull/625

**Comment by jonathanyeung** [comment]
Quick comment about the command name - I'm wondering if the name 'Move To' is too generic. Would it be better if it's called 'Move Selection To' / 'Move Text Selection' / 'Move Selection' ?  With just 'Move To', I would think maybe it could apply to a note, a header, etc. It isn't very clear to me (as an end user) how it relates to:

- Move Note
- Move Header

**Comment by hikchoi** [comment]
> Quick comment about the command name - I'm wondering if the name 'Move To' is too generic. Would it be better if it's called 'Move Selection To' / 'Move Text Selection' / 'Move Selection' ? With just 'Move To', I would think maybe it could apply to a note, a header, etc. It isn't very clear to me (as an end user) how it relates to:
> 
>     * Move Note
> 
>     * Move Header

That does make sense. Long term my goal here is to have `Move To` that is a consolidation of all of them, but since this only does text extractions it would make sense to not have the name generic yet.

**Review by jonathanyeung — CHANGES_REQUESTED** [review]
Just minor comments, looks good!

File: packages/plugin-core/src/commands/NoteLookupCommand.ts [hunk 1]
```diff
@@ -166,6 +166,10 @@ export class NoteLookupCommand
     return this._provider;
   }
 
+  public set provider(provider: ILookupProviderV3 | undefined) {
```
[the remark this answers is not in the document]
Comment (hikchoi) [reply packages/plugin-core/src/commands/NoteLookupCommand.ts:169]: Completely agree. Normally I would go with the typical flow of creating a lookup-derived command (`[[Add a new command that uses lookup|dendron://dendron.docs/pkg.plugin-core.t.lookup.cook#add-a-new-command-that-uses-lookup]]`) and handle it within the new command, but then I will be duplicating the `acceptItem` behavior which I actually don't touch at all. So I went with the route of creating a new provider with the opts I needed, and swapping it out when triggering the already existing lookup command instance.

[the remark this answers is not in the document]
Comment (hikchoi) [reply packages/plugin-core/src/commands/NoteLookupCommand.ts:169]: Lookup is currently only modular until the controller / provider level, so a remedy for this would be splitting out all individual core chunks of logic so that they could be ingredients that we can use on a command-to-command basis

**Comment by hikchoi** [comment]
update:

- `editor.ts` renamed to `EditorUtils.ts` as per convention
- renamed `Move To` to `Move Selection To`
- updated docs to match the rename

**Review by jonathanyeung — APPROVED** [review]
